### ISSUE TITLE
no known conversion from 'float' to 'hip_bfloat16' for 1st argument

### ISSUE THREAD

**Issue by geohot** [tinygrad/tinygrad#3536]
`tiny@tiny5:~/tinygrad$ HSA=1 python3 examples/llama.py  --temperature=0 --count=50 --prompt="Hello." --timing --shard 2 --size 70B --gen 2`

```
using HSA backend                                                                                                                                                                                                                                             
using LLaMA-2-70B model                                                                                                        
ram used:  0.00 GB, layers.0.attention.wq.weight                      :   0%|                                                                                                                                                         | 0/724 [00:00<?, ?it/s]
/tmp/comgr-bf1528/input/<null>:65:43: error: no matching conversion for C-style cast from 'float' to 'hip_bfloat16'
  hip_bfloat16 val0 = (alu3?*(data1+alu2):(hip_bfloat16)(0.0f));                                                                                                                                                                                              
                                          ^~~~~~~~~~~~~~~~~~~~                                                                 
/tmp/comgr-bf1528/input/<null>:10:10: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'float' to 'const hip_bfloat16' for 1st argument
  struct hip_bfloat16 { unsigned short data; };                                                                                                                                                                                                               
         ^                                         
/tmp/comgr-bf1528/input/<null>:10:10: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'float' to 'hip_bfloat16' for 1st argument
  struct hip_bfloat16 { unsigned short data; };                                                                                                                                                                                                               
         ^                                                                                                                     
/tmp/comgr-bf1528/input/<null>:10:10: note: candidate constructor (the implicit default constructor) not viable: requires 0 arguments, but 1 was provided  
```

Looks like it's failing trying to cast the const

**Comment by Zaffer** [comment]
my attempt to resolve the issue #3558

**Comment by Zaffer** [comment]
rocm needs to be updated on tinybox.

appears to be many issues with hip_bfloat16 in the rocm github still open: 
https://github.com/ROCm/ROCm/issues?q=is%3Aissue+is%3Aopen+bfloat16

there are multiple issues with the HIPLanguage class in renderers/cstyle.py:

```python
class HIPLanguage(CStyleLanguage):
    kernel_prefix = "#include <hip/hip_runtime.h>\n"
    # Updated launch bounds definition to align with HIP's expected syntax.
    # Additional HIP includes or definitions can be added here as needed.

    code_for_workitem = {
        "g": lambda x: f"hipBlockIdx_{['x', 'y', 'z'][x]}",
        "l": lambda x: f"hipThreadIdx_{['x', 'y', 'z'][x]}",
        "i": lambda x: f"(hipBlockIdx_{['x', 'y', 'z'][x]} * hipBlockDim_{['x', 'y', 'z'][x]} + hipThreadIdx_{['x', 'y', 'z'][x]})"
    }
    # Updated memory barrier and shared memory prefix
    barrier = "__syncthreads();"
    smem_prefix = "__shared__ "
    float4 = None  # Assuming float4 and similar types are defined in HIP headers
    launch_bounds = True
    uses_ptr_arithmetic = True
    type_map = {dtypes.bfloat16: "__nv_bfloat16"}  # Adjust based on available types in HIP

    def render_kernel(self, function_name: str, kernel: List[str], bufs: List[Tuple[str, DType]], local_size: List[int], uops: List[UOp], prefix=None) -> str:
        # Correcting the kernel function decorator for HIP
        launch_bounds_decorator = f"__attribute__((amdgpu_flat_work_group_size({prod(local_size)}, 1)))"
        kernel_declaration = f"extern \"C\" __global__ {launch_bounds_decorator} void {function_name}("
        kernel_args = ', '.join([f'{self.render_dtype(dtype)}* {name}' for name, dtype in bufs])
        kernel_body = "\n".join(kernel)
        return "\n".join([kernel_prefix, kernel_declaration + kernel_args + ") {", kernel_body, "}"])

```

the underlying causes of this issue seems very advanced.

**Comment by Zaffer** [comment]
all this issues are now resolved in the #3558 PR
